atlasdb-commons/src/main/java/com/palantir/util/IndexEncodingUtils.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.util;
+
+import com.google.common.collect.Maps;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+public final class IndexEncodingUtils {
+    private IndexEncodingUtils() {}
+
+    public static <K, V> Map<Integer, V> encode(List<K> orderedKeys, Map<K, V> values) {
+        return encode(orderedKeys, values, Function.identity());
+    }
+
+    /**
+     * Given an ordered list of keys and a map of keys to values, compute a map that uses indices into the list of keys
+     * instead of actual keys.
+     * Values of the returned map are created using the provided mapping function.
+     *
+     * @param orderedKeys a list of keys in any fixed (!) order. The same list MUST be used for decoding
+     * @param values a map of keys to values. Every key MUST be contained in the list of keys.
+     * @param valueMapper a mapping function to run before putting a value into the result map
+     */
+    public static <K, V, R> Map<Integer, R> encode(List<K> orderedKeys, Map<K, V> values, Function<V, R> valueMapper) {
+        // A linked hash map will give a minor improvement when iterating during serialization
+        Map<Integer, R> res = Maps.newLinkedHashMapWithExpectedSize(values.size());
```

Comment:
this assumes that the result is only iterated through and not heavily used for point lookups